internal/ethapi/api_subscriptions.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package ethapi
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+	"golang.org/x/net/context"
+)
+
+// SubscribeBlocksEvents writes new blocks to the given blocks channel and returns a subscription
+// that can be Unsubscribed when the caller doesn't want to receive blocks any more.
+func (api *PublicBlockChainAPI) SubscribeBlocksEvents(blocks chan *types.Block) *newBlocksSubscription {
+	api.installBlockSubscription <- blocks
+	return &newBlocksSubscription{api, blocks}
+}
+
+// NewBlocksArgs allows the user to specify if the returned block should include transactions and in which format.
+type NewBlocksArgs struct {
+	IncludeTransactions bool `json:"includeTransactions"`
+	TransactionDetails  bool `json:"transactionDetails"`
+}
+
+// NewBlocks send a notification each time a new block is appended to the chain.
+// It accepts an argument which allows the caller to format the block notification.
+func (api *PublicBlockChainAPI) NewBlocks(ctx context.Context, args NewBlocksArgs) (*rpc.Subscription, error) {
```

Comment:
We discussed moving this subscription to the filter system because it also provides new block notifications.